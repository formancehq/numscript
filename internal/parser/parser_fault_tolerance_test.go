package parser_test

import (
	"testing"

	"github.com/formancehq/numscript/internal/parser"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
)

func TestFaultToleranceVarName(t *testing.T) {
	p := parser.Parse(`vars { monetary 42  }`)
	snaps.MatchSnapshot(t, p.Value)
	assert.NotEmpty(t, p.Errors)
}

func TestFaultToleranceSend(t *testing.T) {
	p := parser.Parse(`send `)
	snaps.MatchSnapshot(t, p.Value)
}

func TestFaultToleranceMonetary(t *testing.T) {
	p := parser.Parse(`send [COIN]`)
	snaps.MatchSnapshot(t, p.Value)
}

func TestFaultToleranceNoAddr(t *testing.T) {
	p := parser.Parse(`send  (
  source = {
	@
  }
  destination = @
)`)
	snaps.MatchSnapshot(t, p.Value)
}

func TestFaultToleranceInvalidDest(t *testing.T) {
	p := parser.Parse(`send [COIN 10] (
    source = @a
    destination =
)`)
	snaps.MatchSnapshot(t, p.Value)
}

func TestFaultToleranceInvalidSrcTk(t *testing.T) {
	p := parser.Parse(`send [COIN 10] (
    source = max
    destination = @d
)`)
	snaps.MatchSnapshot(t, p.Value)
}

func TestFaultToleranceTrailingComma(t *testing.T) {
	p := parser.Parse(`set_tx_meta(1, )`)
	snaps.MatchSnapshot(t, p.Value)
}

func TestFaultToleranceDestinationNoRemainingMispelledFrom(t *testing.T) {
	p := parser.Parse(`send [COIN 10] (
		source = @a
		destination = {
			max [COIN 10] from @x
		}
	)
	`)
	snaps.MatchSnapshot(t, p.Value)
}

func TestFaultToleranceIncompleteOrigin(t *testing.T) {
	p := parser.Parse(`
vars {
	asset $a = 
}
	`)
	snaps.MatchSnapshot(t, p.Value)
}
