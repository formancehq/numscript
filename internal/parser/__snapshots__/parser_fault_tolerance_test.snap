
[TestFaultToleranceVarName - 1]
parser.Program{
    Vars: {
        {
            Range: parser.Range{
                Start: parser.Position{Character:7, Line:0},
                End:   parser.Position{Character:18, Line:0},
            },
            Name: (*parser.VariableLiteral)(nil),
            Type: &parser.TypeDecl{
                Range: parser.Range{
                    Start: parser.Position{Character:7, Line:0},
                    End:   parser.Position{Character:15, Line:0},
                },
                Name: "monetary",
            },
            Origin: (*parser.FnCall)(nil),
        },
    },
    Statements: nil,
}
---

[TestFaultToleranceSend - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.SendStatement{
            Range: parser.Range{
                Start: parser.Position{},
                End:   parser.Position{Character:4, Line:0},
            },
            SentValue:   nil,
            Source:      nil,
            Destination: nil,
        },
    },
}
---

[TestFaultToleranceMonetary - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.SendStatement{
            Range: parser.Range{
                Start: parser.Position{},
                End:   parser.Position{Character:11, Line:0},
            },
            SentValue:   nil,
            Source:      nil,
            Destination: nil,
        },
    },
}
---

[TestFaultToleranceNoAddr - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.SendStatement{
            Range: parser.Range{
                Start: parser.Position{},
                End:   parser.Position{Character:1, Line:5},
            },
            SentValue: nil,
            Source:    &parser.SourceInorder{
                Range: parser.Range{
                    Start: parser.Position{Character:11, Line:1},
                    End:   parser.Position{Character:3, Line:3},
                },
                Sources: nil,
            },
            Destination: nil,
        },
    },
}
---

[TestFaultToleranceInvalidDest - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.SendStatement{
            Range: parser.Range{
                Start: parser.Position{},
                End:   parser.Position{Character:1, Line:3},
            },
            SentValue: &parser.SentValueLiteral{
                Range: parser.Range{
                    Start: parser.Position{Character:5, Line:0},
                    End:   parser.Position{Character:14, Line:0},
                },
                Monetary: &parser.MonetaryLiteral{
                    Range: parser.Range{
                        Start: parser.Position{Character:5, Line:0},
                        End:   parser.Position{Character:14, Line:0},
                    },
                    Asset: &parser.AssetLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:6, Line:0},
                            End:   parser.Position{Character:10, Line:0},
                        },
                        Asset: "COIN",
                    },
                    Amount: &parser.NumberLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:11, Line:0},
                            End:   parser.Position{Character:13, Line:0},
                        },
                        Number: 10,
                    },
                },
            },
            Source: &parser.SourceAccount{
                Literal: &parser.AccountLiteral{
                    Range: parser.Range{
                        Start: parser.Position{Character:13, Line:1},
                        End:   parser.Position{Character:15, Line:1},
                    },
                    Name: "a",
                },
            },
            Destination: nil,
        },
    },
}
---

[TestFaultToleranceInvalidSrcTk - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.SendStatement{
            Range: parser.Range{
                Start: parser.Position{},
                End:   parser.Position{Character:1, Line:3},
            },
            SentValue: &parser.SentValueLiteral{
                Range: parser.Range{
                    Start: parser.Position{Character:5, Line:0},
                    End:   parser.Position{Character:14, Line:0},
                },
                Monetary: &parser.MonetaryLiteral{
                    Range: parser.Range{
                        Start: parser.Position{Character:5, Line:0},
                        End:   parser.Position{Character:14, Line:0},
                    },
                    Asset: &parser.AssetLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:6, Line:0},
                            End:   parser.Position{Character:10, Line:0},
                        },
                        Asset: "COIN",
                    },
                    Amount: &parser.NumberLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:11, Line:0},
                            End:   parser.Position{Character:13, Line:0},
                        },
                        Number: 10,
                    },
                },
            },
            Source: &parser.SourceCapped{
                Range: parser.Range{
                    Start: parser.Position{Character:13, Line:1},
                    End:   parser.Position{Character:16, Line:1},
                },
                From: nil,
                Cap:  nil,
            },
            Destination: &parser.DestinationAccount{
                Literal: &parser.AccountLiteral{
                    Range: parser.Range{
                        Start: parser.Position{Character:18, Line:2},
                        End:   parser.Position{Character:20, Line:2},
                    },
                    Name: "d",
                },
            },
        },
    },
}
---

[TestFaultToleranceTrailingComma - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.FnCall{
            Range: parser.Range{
                Start: parser.Position{},
                End:   parser.Position{Character:16, Line:0},
            },
            Caller: &parser.FnCallIdentifier{
                Range: parser.Range{
                    Start: parser.Position{},
                    End:   parser.Position{Character:11, Line:0},
                },
                Name: "set_tx_meta",
            },
            Args: {
                &parser.NumberLiteral{
                    Range: parser.Range{
                        Start: parser.Position{Character:12, Line:0},
                        End:   parser.Position{Character:13, Line:0},
                    },
                    Number: 1,
                },
                nil,
            },
        },
    },
}
---

[TestFaultToleranceDestinationNoRemainingMispelledFrom - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.SendStatement{
            Range: parser.Range{
                Start: parser.Position{},
                End:   parser.Position{Character:2, Line:5},
            },
            SentValue: &parser.SentValueLiteral{
                Range: parser.Range{
                    Start: parser.Position{Character:5, Line:0},
                    End:   parser.Position{Character:14, Line:0},
                },
                Monetary: &parser.MonetaryLiteral{
                    Range: parser.Range{
                        Start: parser.Position{Character:5, Line:0},
                        End:   parser.Position{Character:14, Line:0},
                    },
                    Asset: &parser.AssetLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:6, Line:0},
                            End:   parser.Position{Character:10, Line:0},
                        },
                        Asset: "COIN",
                    },
                    Amount: &parser.NumberLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:11, Line:0},
                            End:   parser.Position{Character:13, Line:0},
                        },
                        Number: 10,
                    },
                },
            },
            Source: &parser.SourceAccount{
                Literal: &parser.AccountLiteral{
                    Range: parser.Range{
                        Start: parser.Position{Character:11, Line:1},
                        End:   parser.Position{Character:13, Line:1},
                    },
                    Name: "a",
                },
            },
            Destination: &parser.DestinationInorder{
                Range: parser.Range{
                    Start: parser.Position{Character:16, Line:2},
                    End:   parser.Position{Character:3, Line:4},
                },
                Clauses: {
                    {
                        Range: parser.Range{
                            Start: parser.Position{Character:3, Line:3},
                            End:   parser.Position{Character:3, Line:4},
                        },
                        Cap: &parser.MonetaryLiteral{
                            Range: parser.Range{
                                Start: parser.Position{Character:7, Line:3},
                                End:   parser.Position{Character:16, Line:3},
                            },
                            Asset: &parser.AssetLiteral{
                                Range: parser.Range{
                                    Start: parser.Position{Character:8, Line:3},
                                    End:   parser.Position{Character:12, Line:3},
                                },
                                Asset: "COIN",
                            },
                            Amount: &parser.NumberLiteral{
                                Range: parser.Range{
                                    Start: parser.Position{Character:13, Line:3},
                                    End:   parser.Position{Character:15, Line:3},
                                },
                                Number: 10,
                            },
                        },
                        To: nil,
                    },
                },
                Remaining: nil,
            },
        },
    },
}
---

[TestFaultToleranceIncompleteOrigin - 1]
parser.Program{
    Vars: {
        {
            Range: parser.Range{
                Start: parser.Position{Character:1, Line:2},
                End:   parser.Position{Character:11, Line:2},
            },
            Name: &parser.VariableLiteral{
                Range: parser.Range{
                    Start: parser.Position{Character:7, Line:2},
                    End:   parser.Position{Character:9, Line:2},
                },
                Name: "a",
            },
            Type: &parser.TypeDecl{
                Range: parser.Range{
                    Start: parser.Position{Character:1, Line:2},
                    End:   parser.Position{Character:6, Line:2},
                },
                Name: "asset",
            },
            Origin: (*parser.FnCall)(nil),
        },
    },
    Statements: nil,
}
---

[TestFaultToleranceIncompleteSave - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.SaveStatement{
            Range: parser.Range{
                Start: parser.Position{Character:0, Line:1},
                End:   parser.Position{Character:4, Line:1},
            },
            SentValue: nil,
            Literal:   nil,
        },
    },
}
---

[TestFaultToleranceIncompleteSave2 - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.SaveStatement{
            Range: parser.Range{
                Start: parser.Position{Character:0, Line:1},
                End:   parser.Position{Character:14, Line:1},
            },
            SentValue: &parser.SentValueLiteral{
                Range: parser.Range{
                    Start: parser.Position{Character:5, Line:1},
                    End:   parser.Position{Character:14, Line:1},
                },
                Monetary: &parser.MonetaryLiteral{
                    Range: parser.Range{
                        Start: parser.Position{Character:5, Line:1},
                        End:   parser.Position{Character:14, Line:1},
                    },
                    Asset: &parser.AssetLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:6, Line:1},
                            End:   parser.Position{Character:10, Line:1},
                        },
                        Asset: "COIN",
                    },
                    Amount: &parser.NumberLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:11, Line:1},
                            End:   parser.Position{Character:13, Line:1},
                        },
                        Number: 10,
                    },
                },
            },
            Literal: nil,
        },
    },
}
---

[TestFaultToleranceIncompleteSave3 - 1]
parser.Program{
    Vars:       nil,
    Statements: {
        &parser.SaveStatement{
            Range: parser.Range{
                Start: parser.Position{Character:0, Line:1},
                End:   parser.Position{Character:19, Line:1},
            },
            SentValue: &parser.SentValueLiteral{
                Range: parser.Range{
                    Start: parser.Position{Character:5, Line:1},
                    End:   parser.Position{Character:14, Line:1},
                },
                Monetary: &parser.MonetaryLiteral{
                    Range: parser.Range{
                        Start: parser.Position{Character:5, Line:1},
                        End:   parser.Position{Character:14, Line:1},
                    },
                    Asset: &parser.AssetLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:6, Line:1},
                            End:   parser.Position{Character:10, Line:1},
                        },
                        Asset: "COIN",
                    },
                    Amount: &parser.NumberLiteral{
                        Range: parser.Range{
                            Start: parser.Position{Character:11, Line:1},
                            End:   parser.Position{Character:13, Line:1},
                        },
                        Number: 10,
                    },
                },
            },
            Literal: nil,
        },
    },
}
---
